### ISSUE TITLE
MIR-based borrow checker complains about rogue elves ("mut elf") when used with arbitrary self types

### ISSUE THREAD

**Issue by shepmaster** [rust-lang/rust#51578]
1.28.0-nightly (2018-06-13 5205ae8bdc2991eecf3b)

```rust
#![feature(nll, arbitrary_self_types)]

use std::rc::Rc;

struct Combine(u8);

impl Combine {
    fn poll(self: Rc<Self>) {
        drop(self);
        self.0 = 42;
    }
}

fn main() {}
```

```
error[E0594]: cannot assign to immutable item
  --> src/main.rs:10:9
   |
10 |         self.0 = 42;
   |         ----^^^^^^^
   |         |
   |         cannot assign through `&`-reference
   |         help: consider changing this to be a mutable reference: `&mut elf`
```

Unsurprisingly, `&mut elf` does not help the code compile.

**Comment by kennytm** [comment]
You don't need arbitrary self types.

```rust
#![feature(nll)]

use std::rc::Rc;

struct Combine(u8);

fn poll(welf: Rc<Combine>) {
    drop(welf);
    welf.0 = 42;
}

fn main() {}
```

**Comment by pnkfelix** [comment]
@pnkfelix wants to double check how this behaves in the context of PR #51275

**Comment by pnkfelix** [comment]
(yes, #51275 appears to fix this; it does not issue the erroneous suggestion in these cases.)

**Comment by nikomatsakis** [comment]
assigning @pnkfelix as they think they have a fix in #51275
